nixos/modules/services/misc/pufferpanel.nix

```diff
@@ -0,0 +1,401 @@
+        '';
+      };
+    };
+
+    environment = lib.mkOption {
+      type = lib.types.attrsOf lib.types.str;
+      default = { };
+      example = lib.literalExpression ''
+        {
+          PUFFER_DAEMON_CONSOLE_BUFFER = "1000";
+          PUFFER_DAEMON_CONSOLE_FORWARD = "true";
+          PUFFER_TOKEN_PUBLIC = "https://panel.example.com/auth/publickey";
+        }
+      '';
+      description = lib.mdDoc ''
+        Environment variables to set for the service. Secrets should be
+        specified using {option}`environmentFile`.
+
+        Refer to the [PufferPanel source code][] for the list of available
+        configuration options. Variable name is an upper-cased configuration
+        entry name with underscores instead of dots, prefixed with `PUFFER_`.
+        For example, `panel.settings.companyName` entry can be set using
+        {env}`PUFFER_PANEL_SETTINGS_COMPANYNAME`.
+
+        [PufferPanel source code]: https://github.com/PufferPanel/PufferPanel/blob/master/config/entries.go
+      '';
+    };
+
+    environmentFiles = lib.mkOption {
+      type = lib.types.listOf lib.types.path;
+      default = [ ];
+      description = lib.mdDoc ''
+        Files to load environment variables from. Loaded variables override
+        values set in {option}`environment`.
+      '';
+    };
+
+    openFirewall = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = lib.mdDoc ''
+        Open ports in the firewall for PufferPanel.
+      '';
+    };
+
+    host = lib.mkOption {
+      type = lib.types.str;
+      default = "";
+      example = "::1";
+      description = lib.mdDoc ''
+        Bind to the specified IPv4 or IPv6 address.
+      '';
+    };
+
+    port = lib.mkOption {
+      type = lib.types.port;
+      default = 8080;
+      description = lib.mdDoc ''
+        Server listen port.
+      '';
+    };
+
+    panel = {
+      enable = lib.mkOption {
+        type = lib.types.bool;
+        default = true;
+        description = lib.mdDoc ''
+          Whether to enable PufferPanel web interface.
+        '';
+      };
+
+      registrationEnabled = lib.mkOption {
+        type = lib.types.nullOr lib.types.bool;
+        default = null;
+        description = lib.mdDoc ''
+          Whether users can register themselves in the web interface.
+
+          To create users when registration is disabled, run
+          {command}`pufferpanel --workDir /var/lib/pufferpanel user add`.
+
+          This option overrides corresponding setting set via web interface
+          after service restart.
+        '';
+      };
+    };
+
+    daemon = {
+      enable = lib.mkOption {
+        type = lib.types.bool;
+        default = true;
+        description = lib.mdDoc ''
+          Whether to enable PufferPanel daemon.
+        '';
+      };
+
+      sftp = {
+        host = lib.mkOption {
+          type = lib.types.str;
+          default = "";
```

Comment:
What means empty string?

Reply:
This is an IP address for [`net.Listen`](https://pkg.go.dev/net#Listen).

>For TCP networks, if the host in the address parameter is empty or a literal unspecified IP address, Listen listens on all available unicast and anycast IP addresses of the local system.

I’ll add this to the option description.